controllers/machineset_controller.go

```diff
@@ -18,31 +18,510 @@ package controllers
+		if metav1.GetControllerOf(machine) == nil {
+			if err := r.adoptOrphan(machineSet, machine); err != nil {
+				klog.Warningf("Failed to adopt Machine %q into MachineSet %q: %v", machine.Name, machineSet.Name, err)
+				r.recorder.Eventf(machineSet, corev1.EventTypeWarning, "FailedAdopt", "Failed to adopt Machine %q: %v", machine.Name, err)
+				continue
+			}
+			klog.Infof("Adopted Machine %q into MachineSet %q", machine.Name, machineSet.Name)
+			r.recorder.Eventf(machineSet, corev1.EventTypeNormal, "SuccessfulAdopt", "Adopted Machine %q", machine.Name)
+		}
+
+		filteredMachines = append(filteredMachines, machine)
+	}
+
+	syncErr := r.syncReplicas(machineSet, filteredMachines)
+
+	ms := machineSet.DeepCopy()
+	newStatus := r.calculateStatus(ms, filteredMachines)
+
+	// Always updates status as machines come up or die.
+	updatedMS, err := updateMachineSetStatus(r.Client, machineSet, newStatus)
+	if err != nil {
+		if syncErr != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to sync machines: %v. failed to update machine set status", syncErr)
+		}
+		return ctrl.Result{}, errors.Wrap(err, "failed to update machine set status")
+	}
+
+	if syncErr != nil {
+		return ctrl.Result{}, errors.Wrapf(syncErr, "failed to sync Machineset replicas")
+	}
+
+	var replicas int32
+	if updatedMS.Spec.Replicas != nil {
+		replicas = *updatedMS.Spec.Replicas
+	}
+
+	// Resync the MachineSet after MinReadySeconds as a last line of defense to guard against clock-skew.
+	// Clock-skew is an issue as it may impact whether an available replica is counted as a ready replica.
+	// A replica is available if the amount of time since last transition exceeds MinReadySeconds.
+	// If there was a clock skew, checking whether the amount of time since last transition to ready state
+	// exceeds MinReadySeconds could be incorrect.
+	// To avoid an available replica stuck in the ready state, we force a reconcile after MinReadySeconds,
+	// at which point it should confirm any available replica to be available.
+	if updatedMS.Spec.MinReadySeconds > 0 &&
+		updatedMS.Status.ReadyReplicas == replicas &&
+		updatedMS.Status.AvailableReplicas != replicas {
+
+		return ctrl.Result{RequeueAfter: time.Duration(updatedMS.Spec.MinReadySeconds) * time.Second}, nil
+	}
+
+	// Quickly rereconcile until the nodes become Ready.
+	if updatedMS.Status.ReadyReplicas != replicas {
+		klog.V(4).Info("Some nodes are not ready yet, requeuing until they are ready")
+		return ctrl.Result{RequeueAfter: 15 * time.Second}, nil
+	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *MachineSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&clusterv1alpha2.MachineSet{}).
-		Complete(r)
+// getCluster reuturns the Cluster associated with the MachineSet, if any.
+func (r *MachineSetReconciler) getCluster(ms *clusterv1.MachineSet) (*clusterv1.Cluster, error) {
```

Comment:
Can we use one of the util helpers here?

Reply:
Done, good catch!